src/artifact_providers/base.ts

```diff
@@ -240,23 +240,23 @@ export abstract class BaseArtifactProvider {
   ): Promise<RemoteArtifact[]> {
     this.logger.debug(`Fetching artifact list for revision \`${revision}\`.`);
     // check the cache first
-    const cached = this.fileListCache[revision];
-    if (cached) {
+    if (this.fileListCache[revision]) {
       this.logger.debug(`Found list in cache.`);
-      return cached;
+    } else {
+      // Cache the promise immediately to cause any subsequent calls during the
+      // fetch to use the pending promise instead of fetching again in parallel
+      this.fileListCache[revision] = this.doListArtifactsForRevision(revision);
     }
 
-    // the data wasn't in the cache, so now we have to go get it
-    let artifacts;
+    let artifacts: RemoteArtifact[];
     try {
-      artifacts = await this.doListArtifactsForRevision(revision);
+      artifacts = (await this.fileListCache[revision]) || [];
```

Comment:
The `|| []` fallback is not really necessary but TypeScript insisted that this may be `undefined`. It clearly cannot read the `if` block above 😀

Reply:
Turns out this was due to the incorrect type def above. Fixing...